finesse/hardware/serial_device.py

```diff
@@ -1,20 +1,43 @@
 """Provides a base class for USB serial devices."""
 from __future__ import annotations
 
+from dataclasses import dataclass
 from typing import Any
 
-from serial import Serial
+from serial import Serial, SerialException
 from serial.tools.list_ports import comports
 
 from finesse.config import BAUDRATES
 from finesse.device_info import DeviceParameter
 
 from .device import AbstractDevice
 
-_serial_ports: list[str] | None = None
+_serial_ports: dict[_USBSerialPortInfo, str] | None = None
 
 
-def _get_usb_serial_ports() -> list[str]:
+@dataclass(frozen=True)
+class _USBSerialPortInfo:
+    """Info to distinguish between USB serial ports."""
+
+    vendor_id: int
+    """USB vendor ID."""
+    product_id: int
+    """USB product ID."""
+    serial_number: str | None
+    """USB serial number."""
+    count: int
+    """How many previous devices match the above parameters."""
+
+    def __str__(self) -> str:
+        out = f"{self.vendor_id:04x}:{self.product_id:04x}"
+        if self.serial_number:
+            out += f" {self.serial_number}"
+        if self.count > 0:
+            out += f" ({self.count+1})"
+        return out
+
+
+def _get_usb_serial_ports() -> dict[_USBSerialPortInfo, str]:
```

Comment:
@dalonsoa This is where it's populated. The result is cached in `_serial_ports` so the ports are only scanned once.